src/library/scala/collection/immutable/ListMap.scala

```diff
@@ -189,8 +217,55 @@ sealed class ListMap[A, +B] extends AbstractMap[A, B]
       else removeInternal(k, cur.next, cur :: acc)
 
     override protected def next: ListMap[A, B1] = ListMap.this
+    @inline private[ListMap] def next0 = next
 
     override def last: (A, B1) = (key, value)
     override def init: ListMap[A, B1] = next
+    override def iterator: Iterator[(A, B1)] = {
+      this.size match {
```

Comment:
How about `private def sizeIsSmallerThan8` to avoid a wasted full iteration for non-small maps?

Reply:
we need to know the size of large maps as well - we use that to size an array and unpack into that to do the iteration
see https://github.com/scala/scala/pull/9220/files#diff-d38adb134d68ac49dc593ad4f2a3c4bd873de22b4c87e1a0607b02893ce4c2baR250